### PR TITLE
[CanonicalizeOSSALifetime] Extend lexical lifetimes to unreachables.

### PR DESCRIPTION
**Pull request by nate-chandler**
When canonicalizing the lifetime of a lexical value, deinit barriers are respected. This is done by walking backwards from lifetime ends and adding encountered deinit barriers to liveness.

Only destroy lifetime ends were walked back from under the assumption that lifetimes would be complete. Without complete OSSA lifetimes, however, it's necessary to also necessary to consider lifetimes that end with unreachables. Unfortunately, we can't simply walk back from those unreachables because there may be instructions which are secretly users of the value being canonicalized (e.g. destroys of `partial_apply`s to which a `begin_borrow` of the value was passed). Such uses don't appear in the use list because lifetime canonicalization expects complete lifetimes and only visits lifetime ends of `begin_borrow`s.

Here, instead, the instructions before the relevant unreachables are added to liveness. In order to determine which unreachables are relevant, it's necessary to have a liveness that includes the original destroys. So a copy of liveness is created and those destroys are added to it.

rdar://115468707

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

File: lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp [hunk 1]
```diff
@@ -249,6 +250,33 @@ void CanonicalizeOSSALifetime::extendLivenessToDeinitBarriers() {
...
+
+  for (auto *end : outsideDestroys) {
+    completeLiveness.updateForUse(end, /*lifetimeEnding*/ true);
+  }
+
+  OSSALifetimeCompletion::visitUnreachableLifetimeEnds(
+      getCurrentDef(), completeLiveness, [&](auto *unreachable) {
+        recordConsumingUser(unreachable);
+        if (auto *previous = unreachable->getPreviousInstruction()) {
+          if (liveness->isInterestingUser(previous) ==
```
Comment (atrick) [inline lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:264]: Let's avoid this bug-prone check in a separate refactoring. Maybe PrunedLiveness needs an extendToNonUse API

**Review by atrick — APPROVED** [review]
Looks good!

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

File: lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:266]: https://github.com/apple/swift/pull/68756

**Pull request merged**